Code/Framework/AzCore/AzCore/Jobs/JobExecutor.cpp

```diff
@@ -0,0 +1,349 @@
+            AZStd::atomic<uint16_t> tail;
+            AZStd::atomic<uint16_t> reserve;
+        };
+
+        // The Job Queue is a lock free 4-priority queue. Its basic operation is as follows:
+        // Each priority level is associated with a different queue, corresponding to the maximum size of a uint16_t.
+        // Each queue is implemented as a ring buffer, and a 64 bit atomic maintains the following state per queue:
+        // - offset to the "head" of the ring, from where we acquire elements
+        // - offset to the "tail" of the ring, which tracks where new elements should be enqueued
+        // - offset to a tail reservation index, which is used to reserve a slot to enqueue elements
+        class JobQueue final
+        {
+        public:
+            // Preallocating upfront allows us to reserve slots to insert jobs without locks.
+            // Each thread allocated by the job manager consumes ~2 MB.
+            constexpr static uint16_t MaxQueueSize = 0xffff;
+            constexpr static uint8_t PriorityLevelCount = static_cast<uint8_t>(JobPriority::PRIORITY_COUNT);
+
+            JobQueue() = default;
+            JobQueue(const JobQueue&) = delete;
+            JobQueue& operator=(const JobQueue&) = delete;
+
+            void Enqueue(TypeErasedJob* job);
+            TypeErasedJob* TryDequeue();
+
+        private:
+            QueueStatus m_status[PriorityLevelCount] = {};
+            TypeErasedJob* m_queues[PriorityLevelCount][MaxQueueSize] = {};
+        };
+
+        void JobQueue::Enqueue(TypeErasedJob* job)
+        {
+            uint8_t priority = job->GetPriorityNumber();
+            QueueStatus& status = m_status[priority];
+
+            AZStd::exponential_backoff backoff;
+            while (true)
+            {
+                uint16_t reserve = status.reserve.load();
+                uint16_t head = status.head.load();
+
+                // Enqueuing is done in two phases because we cannot atomically write the job to the slot we reserve
+                // and simulataneously publish the fact that the slot is now available.
+                if (reserve != head - 1)
+                {
+                    // Try to reserve a slot
+                    if (status.reserve.compare_exchange_weak(reserve, reserve + 1))
+                    {
+                        m_queues[priority][reserve] = job;
+
+                        uint16_t expectedReserve = reserve;
+
+                        // Increment the tail to advertise the new job
+                        while (!status.tail.compare_exchange_weak(expectedReserve, reserve + 1))
+                        {
+                            expectedReserve = reserve;
+                        }
+
+                        return;
+                    }
+
+                    // We failed to reserve a slot, try again
+                }
+                else
+                {
+                    backoff.wait();
+                }
```

Comment:
This can lead to deadlocks if the thread that's failing to queue is also holding on to any resource or work that needs to be completed on the same thread. In practice I've seen this happen when a thread A was trying to queue a job on Thread B but had to wait because Thread B was trying to do the same thing on thread A which is also full.
One possible fix is to have a queue that uses buckets. When a bucket is full a thread creates a new bucket and attempts to cas the pointer to the next bucket to the current bucket and then repeats the process. In practice this is rare enough for the allocation to go largely unnoticed. From there there are a few different approaches to managing the original bucket like leaving it, recycling it or destroying it when the last pop happens on it.
Alternatively don't support `Enqueue` and only have a `TryEnqueue` that's allowed to fail.

Reply:
My initial hunch is that this will be a violation of the job system constraints. Usage of synchronization between jobs that are not marked as explicit edges are undefined behavior. My reasoning here is, the more information we have annotated about how jobs depend on one another, the better our tools can be later.

Reply:
It's not always that clear cut. When I encountered this the chain that happened was this:
- Thread A issued a file request.
- Thread IO started a job.
- The job thread for IO was full which blocked the Thread IO which in turn blocked Thread A.
- The job thread for the IO was done with a previous job and was putting the result in the result queue of Thread A, which was also full.
- Thread A couldn't pump the result queue because it was waiting so there was a deadlock.
It can be argued that these jobs could have been designed better but the point is more that given enough time Murphy's law will provide a deadlock due to filled job queues, so using a queue strategy that supports virtual infinite requests at least makes sure that even the worst job can't deadlock. (Well, at least not on the job system's queue; Murphy's law and all.) When these types of deadlocks happen they're typically not due to direct connection but through some indirection, like the File IO system in this case.